File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/PlatformCoursesAdminLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.core.network.ApiError

object PlatformCoursesAdminLogic {
    const val RBAC_MANAGE_PERMISSION = "global:app:rbac:manage"
    const val DEFAULT_PER_PAGE = 25

    fun adminSettingsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffMobileAdminSettings || features.ffMobileAdminConsole

    fun canManageCourses(permissions: List<String>): Boolean =
        permissions.contains(RBAC_MANAGE_PERMISSION)

    fun shouldShowEntry(features: MobilePlatformFeatures, permissions: List<String>): Boolean =
        !features.ffMobileAdminConsole && features.ffMobileAdminSettings && canManageCourses(permissions)

    fun canView(features: MobilePlatformFeatures, permissions: List<String>): Boolean =
        adminSettingsEnabled(features) && canManageCourses(permissions)

    fun webSettingsPath(): String = "/settings/courses"

    fun courseWebPath(courseCode: String): String = "/courses/$courseCode"

    fun normalizedSearchQuery(query: String): String = query.trim()

    fun shouldSearch(query: String): Boolean = normalizedSearchQuery(query).isNotEmpty()

    fun statusLabel(status: String, active: String, draft: String, archived: String): String =
        when (status.lowercase()) {
            "active" -> active
            "draft" -> draft
            "archived" -> archived
            else -> status
        }

    fun userFacingError(error: Throwable, genericMessage: String): String =
        (error as? ApiError.HttpStatus)?.message?.takeIf { it.isNotEmpty() } ?: genericMessage

    data class MetricDefinition(
        val filter: CoursesListFilter,
        val titleResName: String,
        val hintResName: String?,
        val tableTitleResName: String,
        val tableDescriptionResName: String,
    )

    val METRIC_DEFINITIONS: List<MetricDefinition> = listOf(
        MetricDefinition(
            CoursesListFilter.Created7d,
            "mobile_admin_courses_metric_created7d",
            "mobile_admin_courses_metric_created7d_hint",
            "mobile_admin_courses_metric_created7d_tableTitle",
            "mobile_admin_courses_metric_created7d_tableDescription",
        ),
        MetricDefinition(
            CoursesListFilter.Active,
            "mobile_admin_courses_metric_active",
            "mobile_admin_courses_metric_active_hint",
            "mobile_admin_courses_metric_active_tableTitle",
            "mobile_admin_courses_metric_active_tableDescription",
        ),
        MetricDefinition(
            CoursesListFilter.Draft,
            "mobile_admin_courses_metric_draft",
            "mobile_admin_courses_metric_draft_hint",
            "mobile_admin_courses_metric_draft_tableTitle",
            "mobile_admin_courses_metric_draft_tableDescription",
        ),
        MetricDefinition(
            CoursesListFilter.Total,
            "mobile_admin_courses_metric_total",
            null,
            "mobile_admin_courses_metric_total_tableTitle",
            "mobile_admin_courses_metric_total_tableDescription",
        ),
        MetricDefinition(
            CoursesListFilter.Archived,
            "mobile_admin_courses_metric_archived",
            "mobile_admin_courses_metric_archived_hint",
            "mobile_admin_courses_metric_archived_tableTitle",
            "mobile_admin_courses_metric_archived_tableDescription",
        ),
    )

    fun value(filter: CoursesListFilter, stats: CoursesDashboardStats): Long = when (filter) {
        CoursesListFilter.Created7d -> stats.createdLast7Days
        CoursesListFilter.Active -> stats.activeCourses
        CoursesListFilter.Draft -> stats.draftCourses
        CoursesListFilter.Total -> stats.totalCourses
        CoursesListFilter.Archived -> stats.archivedCourses
    }

    fun toggleFilter(current: CoursesListFilter?, tapped: CoursesListFilter): CoursesListFilter? =
        if (current == tapped) null else tapped

    fun metric(filter: CoursesListFilter): MetricDefinition? =
        METRIC_DEFINITIONS.firstOrNull { it.filter == filter }

    fun formatCount(value: Long): String = "%,d".format(value)
}

File: clients/ios/Lextures/Core/LMS/PlatformCoursesAdminLogic.swift
import Foundation

enum PlatformCoursesAdminLogic {
    static let rbacManagePermission = "global:app:rbac:manage"
    static let defaultPerPage = 25

    static func adminSettingsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffMobileAdminSettings || features.ffMobileAdminConsole
    }

    static func canManageCourses(permissions: [String]) -> Bool {
        permissions.contains(rbacManagePermission)
    }

    static func shouldShowEntry(features: MobilePlatformFeatures, permissions: [String]) -> Bool {
        !features.ffMobileAdminConsole && features.ffMobileAdminSettings && canManageCourses(permissions: permissions)
    }

    static func canView(features: MobilePlatformFeatures, permissions: [String]) -> Bool {
        adminSettingsEnabled(features) && canManageCourses(permissions: permissions)
    }

    static func webSettingsPath() -> String { "/settings/courses" }
    static func courseWebPath(courseCode: String) -> String { "/courses/\(courseCode)" }

    static func normalizedSearchQuery(_ query: String) -> String {
        query.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    static func shouldSearch(_ query: String) -> Bool { !normalizedSearchQuery(query).isEmpty }

    static func statusLabel(_ status: String) -> String {
        switch status.lowercased() {
        case "active": L.text("mobile.admin.courses.status.active")
        case "draft": L.text("mobile.admin.courses.status.draft")
        case "archived": L.text("mobile.admin.courses.status.archived")
        default: status
        }
    }

    static func userFacingError(_ error: Error) -> String {
        if let apiError = error as? APIError, case let .httpStatus(_, message) = apiError,
           let message, !message.isEmpty { return message }
        return L.text("mobile.admin.courses.error")
    }

    struct MetricDefinition: Identifiable, Equatable {
        var id: CoursesListFilter { filter }
        let filter: CoursesListFilter
        let titleKey: String.LocalizationValue
        let hintKey: String.LocalizationValue?
        let tableTitleKey: String.LocalizationValue
        let tableDescriptionKey: String.LocalizationValue
        let systemImage: String
    }

    static let metricDefinitions: [MetricDefinition] = [
        MetricDefinition(
            filter: .created7d,
            titleKey: "mobile.admin.courses.metric.created7d",
            hintKey: "mobile.admin.courses.metric.created7d.hint",
            tableTitleKey: "mobile.admin.courses.metric.created7d.tableTitle",
            tableDescriptionKey: "mobile.admin.courses.metric.created7d.tableDescription",
            systemImage: "book.closed.fill"
        ),
        MetricDefinition(
            filter: .active,
            titleKey: "mobile.admin.courses.metric.active",
            hintKey: "mobile.admin.courses.metric.active.hint",
            tableTitleKey: "mobile.admin.courses.metric.active.tableTitle",
            tableDescriptionKey: "mobile.admin.courses.metric.active.tableDescription",
            systemImage: "book.fill"
        ),
        MetricDefinition(
            filter: .draft,
            titleKey: "mobile.admin.courses.metric.draft",
            hintKey: "mobile.admin.courses.metric.draft.hint",
            tableTitleKey: "mobile.admin.courses.metric.draft.tableTitle",
            tableDescriptionKey: "mobile.admin.courses.metric.draft.tableDescription",
            systemImage: "doc.badge.ellipsis"
        ),
        MetricDefinition(
            filter: .total,
            titleKey: "mobile.admin.courses.metric.total",
            hintKey: nil,
            tableTitleKey: "mobile.admin.courses.metric.total.tableTitle",
            tableDescriptionKey: "mobile.admin.courses.metric.total.tableDescription",
            systemImage: "books.vertical.fill"
        ),
        MetricDefinition(
            filter: .archived,
            titleKey: "mobile.admin.courses.metric.archived",
            hintKey: "mobile.admin.courses.metric.archived.hint",
            tableTitleKey: "mobile.admin.courses.metric.archived.tableTitle",
            tableDescriptionKey: "mobile.admin.courses.metric.archived.tableDescription",
            systemImage: "archivebox.fill"
        ),
    ]

    static func value(for filter: CoursesListFilter, in stats: CoursesDashboardStats) -> Int64 {
        switch filter {
        case .created7d: stats.createdLast7Days
        case .active: stats.activeCourses
        case .draft: stats.draftCourses
        case .total: stats.totalCourses
        case .archived: stats.archivedCourses
        }
    }

    static func toggleFilter(current: CoursesListFilter?, tapped: CoursesListFilter) -> CoursesListFilter? {
        current == tapped ? nil : tapped
    }

    static func metric(for filter: CoursesListFilter) -> MetricDefinition? {
        metricDefinitions.first { $0.filter == filter }
    }
}
